core/modules/replica/MessengerConnector.h

```diff
@@ -0,0 +1,523 @@
+        bool success;
+
+        /// A unique identifier of the request
+        std::string id;
+
+        /// The buffer with a serialized request
+        std::shared_ptr<ProtocolBuffer> requestBufferPtr;
+
+        /// The buffer for receiving responses from a worker server
+        ProtocolBuffer responseBuffer;
+    };
+
+    template <class RESPONSE_TYPE>
+    class Wrapper
+        :   public WrapperBase {
+
+    public:
+
+        typedef std::function<void(std::string const&,
+                                   bool,
+                                   RESPONSE_TYPE const&)> CallbackType;
+
+        // Default construction and copy semantics are prohibited
+
+        Wrapper() = delete;
+        Wrapper(Wrapper const&) = delete;
+        Wrapper& operator=(Wrapper const&) = delete;
+
+        ~Wrapper() override = default;
+
+        /**
+         * The constructor
+         *
+         * @param id                          - a unique identifier of the request
+         * @param requestBufferPtr            - a request serielized into a network buffer
+         * @param responseBufferCapacityBytes - the initial size of the response buffer
+         * @param onFinish                    - an asynchronious callback function called upon
+         *                                      a completion or failure of the operation
+         */
+        Wrapper(std::string const& id,
+                std::shared_ptr<ProtocolBuffer> const& requestBufferPtr,
+                size_t responseBufferCapacityBytes,
+                CallbackType onFinish)
+            :   WrapperBase(id,
+                            requestBufferPtr,
+                            responseBufferCapacityBytes),
+                _onFinish(onFinish) {
+        }
+
+        /**
+         * @see WrapperBase::parseResponseAndNotify
+         */
+        void parseAndNotify() override {
+            RESPONSE_TYPE response;
+            if (success) {
+                try {
+                    responseBuffer.parse(response, responseBuffer.size());
+                } catch(std::runtime_error const& ex) {
+
+                    // The message is corrupt. Google Protobuf will report an error
+                    // of the following kind:
+                    //
+                    //   [libprotobuf ERROR google/protobuf/message_lite.cc:123] Can't parse message of type ...
+                    //
+                    success = false;
+                }
+            }
+            _onFinish(id, success, response);
+        }
+
+    private:
+
+        /// The collback fnction to be called upon the completion of the transaction
+        CallbackType _onFinish;
+    };
+
+    // Default construction and copy semantics are prohibited
+
+    MessengerConnector() = delete;
+    MessengerConnector(MessengerConnector const&) = delete;
+    MessengerConnector& operator=(MessengerConnector const&) = delete;
+
+    ~MessengerConnector() = default;
```

Comment:
IMO it's weird to have these below the nested classes. Is there code standard or convention for it? 
I think it would be best to move out the nested classes (as noted above). 2nd best would be to put these near the top of the class declaration.

Reply:
Fixed